docs/proposed/psa-driver-interface.md

```diff
@@ -0,0 +1,627 @@
+
+* `"prefix"` (mandatory, string). This must be a valid prefix for a C identifier. All the types and functions provided by the driver have a name that starts with this prefix unless overridden with a `"name"` element in the applicable capability as described below.
+* `"type"` (mandatory, string). One of `"transparent"` or `"opaque"`.
+* `"headers"` (optional, array of strings). A list of header files. These header files must define the types provided by the driver and may declare the functions provided by the driver. They may include other PSA headers and standard headers of the platform. Whether they may include other headers is implementation-specific. If omitted, the list of headers is empty.
+* `"capabilities"` (mandatory, array of [capabilities](#driver-description-capability)).
+A list of **capabilities**. Each capability describes a family of functions that the driver implements for a certain class of cryptographic mechanisms.
+* `"key_context"` (not permitted for transparent drivers, mandatory for opaque drivers): information about the [representation of keys](#key-format-for-opaque-drivers).
+* `"persistent_state_size"` (not permitted for transparent drivers, optional for opaque drivers, integer or string). The size in bytes of the [persistent state of the driver](#opaque-driver-persistent-state). This may be either a non-negative integer or a C constant expression of type `size_t`.
+* `"location"` (not permitted for transparent drivers, optional for opaque drivers, integer or string). The location value for which this driver is invoked. In other words, this determines the lifetimes for which the driver is invoked. This may be either a non-negative integer or a C constant expression of type `psa_key_location_t`.
+
+#### Driver description capability
+
+A capability declares a family of functions that the driver implements for a certain class of cryptographic mechanisms. The capability specifies which key types and algorithms are covered and the names of the types and functions that implement it.
+
+A capability is a JSON object containing the following properties:
+
+* `"functions"` (optional, list of strings). Each element is the name of a [driver function](#driver-functions) or driver function family. If specified, the core will invoke this capability of the driver only when performing one of the specified operations. If omitted, the `"algorithms"` property is mandatory and the core will invoke this capability of the driver for all operations that are applicable to the specified algorithms. The driver must implement all the specified or implied functions, as well as the types if applicable.
+* `"algorithms"` (optional, list of strings). Each element is an [algorithm specification](#algorithm-specifications). If specified, the core will invoke this capability of the driver only when performing one of the specified algorithms. If omitted, the core will invoke this capability for all applicable algorithms.
```

Comment:
Not just as implied by `"functions"`, but also by `"key_types"` and the other properties in this section. I'm not sure how to make this clearer. Does this really need a formal specification?

Reply:
I guess that we would like this to be clear enough for two different implementations to be as close as possible. With my proposals I was trying to explain more the relationships between the properties and it seems I didn't get it right. Those dependencies/relationships do not seem to be easy to explain thus this maybe deserves a specific section with examples (just a thought).

Reply:
Revisiting this, the case where `"entry_points"` was omitted was not well-defined. I decided to remove it not only because it complicates the semantics of capabilities, but also because it makes drivers more fragile: for example, if a new compound mechanism such as multipart-hash-and-sign is introduced, a driver that specify `{"algorithms": "PSA_ALG_RSA_PSS(PSA_ALG_SHA_256)"}` but only implements the original entry points would no longer work.

I've also added a subsection that explains the semantics of capabilities, to make things clearer.